cep-20.2.md

```diff
@@ -0,0 +1,788 @@
+  # the interpreter to use for the script
+  interpreter: str  # defaults to bash on UNIX and cmd.exe on Windows
+  # the script environment. You can use Jinja to pass through environment variables
+  # with the `env` key (`${{ env.get("MYVAR") }}`).
+  env: {str: str}
+  # secrets that are set as env variables but never shown in the logs or the environment
+  secrets: [str]
+  # The file to use as the script. Automatically adds `bat` or `sh` to the filename
+  # on Windows or UNIX respectively (if no file extension is given).
+  file: str  # build.sh  
+  # A string or list of strings that is the script contents (mutually exclusive with `file`)
+  content: str | [str]
+```
+
+## Source section
+
+```yaml
+source: [SourceElement]
+```
+
+where the different source elements are defined as follows.
+
+### URL source
+
+```yaml
+# url pointing to the source tar.gz|zip|tar.bz2|... (this can be a list of mirrors that point to the same file)
+url: url | [url]
+# destination folder in work directory
+target_directory: path
+# rename the downloaded file to this name
+file_name: str
+# hash of the file
+sha256: hex string
+# legacy md5 sum of the file (test both, prefer sha256)
+md5: hex string
+# relative path from recipe file
+patches: [path]
+```
+
+### Local source
+
+A path can be either:
+
+- directory ("../bla")
+- a path to an archive ("../bla.tar.gz")
+- a path to a file ("../bla.txt")
+
+```yaml
+# file, absolute or relative to recipe file
+path: path
+# if there is a gitignore, adhere to it (or not)
+use_gitignore: bool (defaults to true)
+# destination folder
+target_directory: path
+# rename the downloaded file to this name
+file_name: str
+# absolute or relative path from recipe file
+patches: [path]
+```
+
+### Git source
+
+```yaml
+# URL to the git repository or path to local git repository
+git: url | path
+# the following 3 keys are mutually exclusive (branch, tag, and rev)
+# branch to checkout to
+branch: string
+# tag to checkout to
+tag: string
+# revision to checkout to (hash or ref)
+rev: string
+# depth of the git clone (mutually exclusive with rev)
+depth: signed integer (defaults to -1 -> not shallow)
+# should this use git-lfs?
+lfs: bool (defaults to false)
+# destination folder in work directory
+target_directory: path
+# absolute or relative path from recipe file
+patches: [path]
+```
+
+### Removed source definitions
+
+SVN and HG (mercury) source definitions are removed as they are not relevant anymore.
+
+## Requirements section
+
+```yaml
+requirements:
+  # build time dependencies, in the build_platform architecture
+  build: [MatchSpec]
+  # dependencies to link against, in the target_platform architecture
+  host: [MatchSpec]
+  # the section below is copied into the index.json and required at package installation
+  run: [MatchSpec]
+  # constrain optional packages
+  run_constrained: [MatchSpec]
+  # the run exports of this package
+  run_exports: [MatchSpec] OR RunExports
+```
+
+#### `RunExports` section
```

Comment:
Has it been considered to remove `run_exports`? It would be nicer, IMHO, if the information for usage of a host or build dependency was captured somehow, and then it was up to the author of the downstream recipe to correctly specify the run dependencies manually. More explicit, easier to debug dependency conflicts, etc.

Reply:
yes, once the new format is done all future changes will be much easier because we can properly lint the YAML and do automatic upgrades. The initial one is the hardest.

There is no problem changing the recipe spec in the future.

Reply:
Thanks, so basically the main thing we're tackling in this chain of CEPs is to remove the semantic comments/jinja flow-control/other stuff that goes away from pure YAML with a few jinja functions and variable substitution?

Reply:
Yeah I would also expect _a lot_ of opposition. Would love to see a CEP with a better idea than run exports though! I am also open to improve the words (weak, strong, etc.) but haven't come up wtih anything better yet (they would still "render" to the same run_exports.json file as today, just different name in the recipe).

Reply:
@danpetry you may not have experienced the compatibility hell before run_exports. The whole idea is that the people who maintain a library (or the package for it) are the best ones to understand how its compatibility changes. If you remove the idea of run_exports, you are instead imposing that knowledge on every consumer of the package. It requires a LOT more knowledge from a lot more people.

People at NVIDIA have asked if we could be more descriptive about where a particular dependency in the rendered recipe came from. This might help with your confusion without sacrificing the centralization of compatibility knowledge.

For example, some verbose mode could output something when a dependency's run_exports are being taken into account. This could just be log output, or it could somehow annotate a rendered version of the recipe.